docs/developer/best_practices/commit.rst

```diff
@@ -0,0 +1,372 @@
+Commit Best Practices
+---------------------
+
+How to split a change into commits
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Why do we need to split changes into commits
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This has several advantages amongst which are:
+
+- small commits are easier to review (a large pull request correctly divided
+  into commits is easier/faster to review than a medium-sized one with less
+  thought-out division)
+
+- simple commits are easier to revert (`e866b01f0553 <https://github.com/scality/metalk8s/commit/e866b01f05535925e80da20aca00417904422433>`_/`8208a170ac66 <https://github.com/scality/metalk8s/commit/8208a170ac66912ace018bcd00c058ad214d169b>`_)/cherry-pick
+  (`Pull request #1641 <https://github.com/scality/metalk8s/pull/1641>`_)
+
+- when looking for a regression (e.g. using ``git bisect``) it is easier to
+  find the root cause
+
+- make ``git log`` and ``git blame`` way more useful
+
+Examples
+^^^^^^^^
+
+The golden rule to create good commits is to ensure that the commit is atomic.
+A commit is atomic when:
+
+- it can be reverted without breaking the product
```

Comment:
Maybe that's a property we *want*, but it's definitely harder to achieve than having the ability to `checkout` any commit and have working code.

To me, `revert` ing a commit is a dangerous operation, and will always be: you never know what code was introduced after this commit, and if it relies on the contents you're removing. We are not reverting code often, and when we do, it's often one of two cases:
- disabling a commit in forward-merges (e.g. some implementation only needed for K8s versions before 1.15)
- removing a commit that works around a limitation of a dependency of our product, that we want to update

IMHO, `revert` is rarely used to remove an error unvoluntarily introduced. We have tests and reviews to prevent bugs from landing in the code, and shouldn't optimize our workflows to "patch" our mistakes after the act.

So, atomicity may be defined in terms of this ability to `revert`, but I'd rather focus on having the ability to **use** the code in any given point of history rather than being able to remove parts of it (not that usual of an operation). Furthermore, when `bisect`ing, ability to `revert` is irrelevant: the most important is that the code should not be broken at all, and we're trying to find one commit that is actually breaking something (maybe due to some history manipulation, or considering another, newer, view of _what is a correct state_).

Reply:
> use the code in any given point of history 

This seems pretty hard to achieve, because most of the code we add depends on already existing code, no?
So if some required code is renamed or moved, then past that point you won't be able to "use it" as-is.

---

Side note: I don't think we will ever come up with a precise definition for an atomic (reversible, logical, you name it) commit that'll stand a reality check (and that's why I didn't try to define it in my first version :stuck_out_tongue: ).

I think it's intrinsically ambiguous and everyone will see it a bit different.
Is it a big deal?
I'm not sure. Most of our current problem are regarding commit message, split is mostly OK.

/me currently wondering if he should go back to an intendedly open "definition"

Reply:
For the record, the concept of "logical commit" (without more definition) is coming from [here](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), which was one of the three main sources that have influenced me.

Reply:
> > use the code in any given point of history

> This seems pretty hard to achieve, because most of the code we add depends on already existing code, no?
> So if some required code is renamed or moved, then past that point you won't be able to "use it" as-is.

That's not what I meant. My point is, any **merged** commit should work as is, i.e. for any commit in `development/x.y`, I can ` git checkout` said commit and have working code.

This implies:
- reordering is a dangerous operation
- `cherry-pick`ing and `revert`ing are dangerous as well

---

I like the "open" definition from your link, that would suit me.

Reply:
> That's not what I meant. My point is, any merged commit should work as is, i.e. for any commit in development/x.y, I can git checkout said commit and have working code.

Ok, I agree with that.

> I like the "open" definition from your link, that would suit me.

Then I'll revert back to my "open" denomination. And if the need for a more precise definition occurs, we can then edit this document (it's not set in stone).